### DISCUSSION TITLE
Persistent timeouts

### DISCUSSION THREAD

**Discussion by MikeBishop** [timdorr/tesla-api#344] [Q&A]
My app has recently been seeing large timeouts (5s timeouts used to be sufficient, now I'm not getting answers at all) with the auth endpoint; both refresh and OAuth scraping are slow.  I assume this is the WAF again.  Anyone know what headers need to be sent to make the WAF happy?

**Comment by MikeBishop** [comment]
Stranger still, I have two different environments where the timeouts happen, but other environments where the same request (I think) gets a successful response.  Something is different, but I can't tell what.

**Reply by hemang249** [reply]
Make sure in your request you are not sending any headers. Only headers you should be sending are Content-Type and Authorization when required.

**Reply by zt4** [reply]
> Make sure in your request you are not sending any headers. Only headers you should be sending are Content-Type and Authorization when required.

That worked for me. I had to stop sending a user-agent header to avoid constant timeouts. Thank you so much!

**Reply by MikeBishop** [reply]
That might be my underlying issue -- most libraries I can find inject their own UA if none is provided, so keeping it from being added is the tricky part.

**Reply by hemang249** [reply]
what is the framework / language you are working with ? I have not had any cases with Python [Requests] or Node js  [node-fetch] . Not sure about about others

**Reply by MikeBishop** [reply]
I'm seeing it with both node-fetch and axios.  I've submitted a PR to axios to support omitting the UA header.

**Reply by laemen** [reply]
> Make sure in your request you are not sending any headers. Only headers you should be sending are Content-Type and Authorization when required.

Thanks that worked for me too!

**Reply by MikeBishop** [reply]
Hm.  Even having the version of axios that can omit sending a user-agent, a very simple test program fails to get a response to the initial OAuth URL.  I'm stymied.

**Reply by hemang249** [reply]
would you mind sharing your request code ? So i can have a better understanding of how you are making the request.

Also one thing to note is too many failed attempts will result in tesla blocking you for a while and you will get back a  403 Forbidden status.

**Reply by hemang249** [reply]
Here is a minimal working code using axios ^0.21.1  for step 1 : Getting the auth form

Note : It would be better to wrap the code in a loop as the first attempt may or may not always succeed and may require a retry

```
const randomstring = require("randomstring");
const crypto = require("crypto");
const base64url = require("base64url");
const axios = require("axios").default;

const code_verifier = randomstring.generate(86);

const base64Digest = crypto
  .createHash("sha256")
  .update(code_verifier)
  .digest("base64");

let audience = "";
let client_id = "ownerapi";
let code_challenge_method = "S256";
let locale = "en";
let prompt = "login";
let redirect_uri = "https://auth.tesla.com/void/callback";
let response_type = "code";
let scope = "openid email offline_access";
let state = base64url.fromBase64(base64Digest);
let code_challenge = base64url.fromBase64(base64Digest);

axios
  .get(
    `https://auth.tesla.com/oauth2/v3/authorize?audience=${audience}&client_id=${client_id}&code_challenge_method=${code_challenge_method}&locale=${locale}&prompt=${prompt}&redirect_uri=${redirect_uri}&response_type=${response_type}&scope=${scope}&state=${state}&code_challenge=${code_challenge}`,
    {
      headers: {},
    }
  )
  .then((res) => {
    console.log(res);
    // do other steps
  })
  .catch((err) => console.log(err));

```

**Reply by MikeBishop** [reply]
I found some differences in URL generation between your code and what I'm working with, and resolving those got it back to working on my desktop.  (I wasn't generating locale, prompt, or audience; some things were being URL-encoded before passing to axios, whereas yours weren't.)

But the [same code that works](https://gist.github.com/MikeBishop/4bb7c949716f0e5ad65f3e4b7485c84c) on my desktop under WSL fails on my Raspberry Pi.  There must be an impact of the environment difference, but I can't figure it out.  I've got the same version of Node, same version of Axios, etc.

(Incidentally, if you pass `headers: {}` into Axios, your UA string will be `axios/(version number)`, so it's not the absence of User-Agent.)

**Reply by MikeBishop** [reply]
On the Pi, your code as-is never resolves the Promise; if I add a 5s timeout to the Axios call, it bails after 5 seconds.

**Reply by hemang249** [reply]
Hm.. I'm unsure about why it stops working on a PI. In my production aws / local environment I am currently using a py based script. And on both it seems to work fine. I will look more into an implementation using node js and see if I can get it to work.

Give this script a go meanwhile, if it works for you 

https://github.com/enode-engineering/tesla-oauth2/blob/main/tesla.py

**Reply by MikeBishop** [reply]
I suspect that it would, but the project I'm working in is constrained to node.  I appreciate your willingness to look into it -- I'm stumped at this point.
